File: baekjoon/삼성SW역량테스트/14502_연구소.cpp
/*
[문제]
크기가 NxM인 연구소가 있다.
0: 빈 칸, 1: 벽, 2: 바이러스가 있는 곳이다.
바이러스는 상하좌우로 인접한 빈 칸으로 모두 퍼져나갈 수 있다.

[출력]
벽 3개를 세워 얻을 수 있는 안전 영역 크기의 최댓값을 구하라.
*/


#include<cstdio>
#include<cstring>
#include<vector>
#include<queue>
#include<algorithm>
#define MAX 8
using namespace std;
struct node {
    int y, x;
};
int N, M;
int TotalCnt, ans;
int map[MAX][MAX];
int visited[MAX][MAX];
int dy[] = { -1,1,0,0 };
int dx[] = { 0,0,-1,1 };
vector<node> virus;

int bfs() {
    int cnt = 0;
    queue<node> q;
    for (auto next : virus) {
        q.push(next);
        visited[next.y][next.x] = 1;
    }

    while (!q.empty()) {
        int y = q.front().y, x = q.front().x; q.pop();
        for (int k = 0; k < 4; k++) {
            int ny = y + dy[k], nx = x + dx[k];
            if (ny < 0 || ny >= N || nx < 0 || nx >= M ||
                map[ny][nx]>0 || visited[ny][nx]) continue;
            q.push({ ny,nx });
            visited[ny][nx] = 1;
            cnt++;
        }
    }
    return cnt;
}

void choice_wall(int y, int x, int cnt) {
    if (cnt == 3) { // 벽이 3개가 되면
        memset(visited, 0, sizeof(visited));
        int cnt = bfs(); // 바이러스 퍼뜨리기
        // 안전 영역 최대 크기 구하기
        ans = max(ans, TotalCnt - cnt);
        return;
    }
    while (y < N) {
        if (map[y][x] == 0) {
            // 벽세우기
            map[y][x] = 1;
            choice_wall(y, x, cnt + 1);
            map[y][x] = 0;
        }
        x++;
        if (x >= M) {
            y++;
            x = 0;
        }
    }
}

int main() {
    scanf("%d %d", &N, &M);
    for (int i = 0; i < N; i++) {
        for (int j = 0; j < M; j++) {
            scanf("%d", &map[i][j]);
            if (map[i][j] == 0) TotalCnt++;
            else if (map[i][j] == 2) virus.push_back({ i,j });
        }
    }
    TotalCnt -= 3;
    choice_wall(0, 0, 0);
    printf("%d", ans);
    return 0;
}

File: baekjoon/삼성SW역량테스트/14502_연구소.py
#include<iostream>
#include<algorithm>
#include<queue>
#define MAX 8
using namespace std;

int dy[4] = { -1,1,0,0 };
int dx[4] = { 0,0,-1,1 };
int N, M;
int lab[MAX][MAX];
int temp[MAX][MAX];
int result=0;

void bfs() {
	int afterSpread[MAX][MAX];
	copy(&temp[0][0], &temp[0][0] + MAX * MAX, &afterSpread[0][0]);

	queue<pair<int, int>> q;
	for (int i = 0; i < N; i++) {
		for (int j = 0; j < M; j++) {
			if (afterSpread[i][j] == 2) {
				q.push(make_pair(i, j));
			}
		}
	}

	while (!q.empty()) {
		int y = q.front().first;
		int x = q.front().second;
		q.pop();

		for (int i = 0; i < 4; i++) {
			int ny = y + dy[i];
			int nx = x + dx[i];

			if (0 <= ny && ny < N && 0 <= nx && nx < M) {
				if (afterSpread[ny][nx] == 0) {
					afterSpread[ny][nx] = 2;
					q.push(make_pair(ny, nx));
				}
			}

		}
	}

	int empty = 0;
	for (int i = 0; i < N; i++) {
		for (int j = 0; j < M; j++) {
			if (afterSpread[i][j] == 0) empty++;
		}
	}
	result = max(result, empty);
}

void make_wall(int r, int c, int n) {
	if (n == 3) {
		bfs();
		return;
	}
	while (r < N) {
		c++;
		if (c == M) {
			c = 0;
			r++;
			if (r == N)
				return;
		}
		if (temp[r][c] == 0) {
			temp[r][c] = 1;
			make_wall(r,c,n + 1);
			temp[r][c] = 0;
		}
	}
}

 int main() {
	cin >> N >> M;
	for (int i = 0; i < N; i++) {
		for (int j = 0; j < M; j++) {
			cin >> lab[i][j];
		}
	}
	copy(&lab[0][0], &lab[0][0] + MAX * MAX, &temp[0][0]);
	for (int i = 0; i < N; i++) {
		for (int j = 0; j < M; j++) {
			if (temp[i][j] == 0) {
				temp[i][j] = 1;
				make_wall(i,j,1);
				temp[i][j] = 0;
			}
		}
	}

	cout << result;
	return 0;
}
